tests/test_graph.py

```diff
@@ -108,6 +110,185 @@ def test_add_dependency_strict_node_dependencies():
+    e_sig = inspect.signature(tests.resources.incompatible_input_types.e)
+    f_sig = inspect.signature(tests.resources.incompatible_input_types.f)
+
+    nodes = {
+        "e": node.Node.from_fn(tests.resources.incompatible_input_types.e),
+        "f": node.Node.from_fn(tests.resources.incompatible_input_types.f),
+    }
+    nodes["e"]._originating_functions = (tests.resources.incompatible_input_types.e,)
+    nodes["f"]._originating_functions = (tests.resources.incompatible_input_types.f,)
+    param_name = "d"
+
+    # this adds 'a' to nodes
+    graph.add_dependency(
+        nodes["e"],
+        "e",
+        nodes,
+        param_name,
+        e_sig.parameters[param_name].annotation,
+        base.SimplePythonDataFrameGraphAdapter(),
+    )
+
+    assert "d" in nodes
+
+    # adding dependency of c on a should fail because the types are incompatible
+    with pytest.raises(ValueError):
+        graph.add_dependency(
+            nodes["e"],
+            "e",
+            nodes,
+            param_name,
+            f_sig.parameters[param_name].annotation,
+            base.SimplePythonDataFrameGraphAdapter(),
+        )
+
+
+def test_add_dependency_input_nodes_compatible_types():
+    """Tests that if functions request an input that we correctly accept compatible types."""
+    b_sig = inspect.signature(tests.resources.compatible_input_types.b)
+    c_sig = inspect.signature(tests.resources.compatible_input_types.c)
+    d_sig = inspect.signature(tests.resources.compatible_input_types.d)
+
+    nodes = {
+        "b": node.Node.from_fn(tests.resources.compatible_input_types.b),
+        "c": node.Node.from_fn(tests.resources.compatible_input_types.c),
+        "d": node.Node.from_fn(tests.resources.compatible_input_types.d),
+    }
+    nodes["b"]._originating_functions = (tests.resources.compatible_input_types.b,)
+    nodes["c"]._originating_functions = (tests.resources.compatible_input_types.c,)
+    nodes["d"]._originating_functions = (tests.resources.compatible_input_types.d,)
+    # what we want to add
+    param_name = "a"
+
+    # this adds 'a' to nodes
+    graph.add_dependency(
+        nodes["b"],
+        "b",
+        nodes,
+        param_name,
+        b_sig.parameters[param_name].annotation,
+        base.SimplePythonDataFrameGraphAdapter(),
+    )
+
+    assert "a" in nodes
+
+    # this adds 'a' to 'c' as well.
+    graph.add_dependency(
+        nodes["c"],
+        "c",
+        nodes,
+        param_name,
+        c_sig.parameters[param_name].annotation,
+        base.SimplePythonDataFrameGraphAdapter(),
+    )
+
+    # test that we shrink the type to the tighter type
+    assert nodes["a"].type == str
+
+    graph.add_dependency(
+        nodes["d"],
+        "d",
+        nodes,
+        param_name,
+        d_sig.parameters[param_name].annotation,
+        base.SimplePythonDataFrameGraphAdapter(),
+    )
+
+
+def test_add_dependency_input_nodes_compatible_types_order_check():
```

Comment:
Not convinced this test is necessary -- a lot of specific stuff for a strange edge case? Basically copied from the previous? Is there something more specific to test? Otherwise, yeah, comment more -- why would order matter/break it. Could also parameterize (E.G. `reversed`) to share code with the previous.

Reply:
added comment -- I don't see value in refactoring the code because it's pretty straightforward and only has two cases.